rln/src/lib.rs

```diff
@@ -87,54 +86,64 @@ mod test {
+        let TREE_HEIGHT = 16;
+        let leafIndex = 3;
 
-        // generate identity
+        // Generate identity
+        // We follow zk-kit approach for identity generation
         let id = Identity::from_seed(b"hello");
+        let identity_secret = poseidon_hash(&vec![id.trapdoor, id.nullifier]);
+        let id_commitment = poseidon_hash(&vec![identity_secret]);
 
-        // generate merkle tree
-        let mut tree = PoseidonTree::new(21, leaf);
-        tree.set(0, id.commitment().into());
+        //// generate merkle tree
+        let leaf = Field::from(0);
+        let mut tree = PoseidonTree::new(TREE_HEIGHT, leaf);
+        tree.set(leafIndex, id_commitment.into());
 
-        let merkle_proof = tree.proof(0).expect("proof should exist");
-        let root = tree.root().into();
+        let merkle_proof = tree.proof(leafIndex).expect("proof should exist");
 
-        println!("Root: {:#}", root);
-        println!("Merkle proof: {:#?}", merkle_proof);
+        let signal = b"hey hey";
+        let x = hash_to_field(signal);
 
-        // change signal_hash and external_nullifier_hash here
-        let signal_hash = hash_to_field(b"xxx");
-        let external_nullifier_hash = hash_to_field(b"appId");
+        // We set the remaining values to random ones
+        let epoch = hash_to_field(b"test-epoch");
+        let rln_identifier = hash_to_field(b"test-rln-identifier");
 
-        let nullifier_hash = generate_nullifier_hash(&id, external_nullifier_hash);
-
-        let proof =
-            generate_proof(&id, &merkle_proof, external_nullifier_hash, signal_hash).unwrap();
-
-        println!("Proof: {:#?}", proof);
-
-        // TODO Make this test pass
-        //
-        // Currently fails at:
-        // thread 'test::test_end_to_end' panicked at 'called `Result::unwrap()`
-        // on an `Err` value: SynthesisError(MalformedVerifyingKey)',
-        // rln/src/lib.rs:62:84
-        //
-        // Not sure why this is MalformedVerifyingKey, though the proof is
-        // likely incorrect with wrong fields in protocol.rs
-        //
-        // Indeed:
-        // if (public_inputs.len() + 1) != pvk.vk.gamma_abc_g1.len() {
-        let success = verify_proof(
-            root,
-            nullifier_hash,
-            signal_hash,
-            external_nullifier_hash,
-            &proof,
-        )
-        .unwrap();
+        let rlnWitness: RLNWitnessInput =
+            initRLNWitnessFromValues(identity_secret, &merkle_proof, x, epoch, rln_identifier);
+
+        println!("rlnWitness: {:#?}", rlnWitness);
+
+        // We generate all relevant keys
+        let provingKey = &ZKEY();
+        let verificationKey = &VK();
+        let mut builder = CIRCOM();
+
+        // Let's generate a zkSNARK proof
+        let (proof, inputs) = generate_proof(builder, provingKey, rlnWitness).unwrap();
+
+        // Let's verify the proof
+        let success = verify_proof(verificationKey, proof, inputs).unwrap();
```

Comment:
Why did the API change here?

Reply:
I thought is better to standardize the APIs within each submodule in zerokit, abstracting the required inputs. For generating proof: the circom builder, a proving key and a witness. For verifying proofs the verification key, the proof and the circuit inputs to be verified.

In this way, for each submodule we need only to take care about generating witnesses, while the function signatures (modulo the type of the witness of course) will remain all the same (and do not depend on the module used). 

We can of course discuss the best approach here, not hard to change back.

Reply:
OK, agree that'd be better, as long as we make the public API and C API match roughly what already exists, so should be fine